### PR TITLE
Profile

### PR DESCRIPTION
**Pull request by anshul2401**
Password is getting updated but their is problem in updating username. Please check it once.

### CODE REVIEW

**Comment by harsimranmaan** [comment]
I tried testing this. The profile page hanged. Did not troubleshoot yet. Will take a look again tomorrow

**Comment by anshul2401** [comment]
Fixed update password issue.

**Comment by harsimranmaan** [comment]
`flutter analyze fails`

File: lib/screens/profile.dart [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+                                }
+                                print(_password2);
+                                if (value != _password2) {
+                                  return ShipantherLocalizations.of(context)
+                                      .passwordUpdateError;
+                                }
+                                return null;
+                              },
+                              onSaved: (val) =>
+                                  setState(() => _confirmPassword = val),
```
Comment (harsimranmaan) [inline lib/screens/profile.dart:182]: `_confirmPassword` is never used

File: lib/bloc/auth/auth_bloc.dart [hunk 2]
```diff
@@ -80,5 +80,17 @@ class AuthBloc extends Bloc<AuthEvent, AuthState> {
       await event.user.sendEmailVerification();
       yield AuthEmailResent(event.user);
     }
+    if (event is UpdatePassword) {
+      var user = _authRepository.loggedInUser();
+
+      await user.updatePassword(event.password);
+
+      yield AuthInitial();
```
Comment (harsimranmaan) [inline lib/bloc/auth/auth_bloc.dart:88]: Why return the inital state? Shouldn't this be a new state?

File: lib/widgets/shipanther_scaffold.dart [hunk 3]
```diff
@@ -43,7 +44,40 @@ class ShipantherScaffold extends StatelessWidget {
           // Important: Remove any padding from the ListView.
           padding: EdgeInsets.zero,
           children: <Widget>[
-                _createHeader(user),
+                // _createHeader(),
+                UserAccountsDrawerHeader(
```
Comment (harsimranmaan) [inline lib/widgets/shipanther_scaffold.dart:48]: This should go in `_createHeader`. It is important to keep changes from main when rebasing and add on top of it.

File: lib/widgets/shipanther_scaffold.dart [hunk 4]
```diff
@@ -153,6 +187,17 @@ List<Widget> drawerItemsFor(BuildContext context, api.User user) {
       ),
     );
   }
+  // widgets.add(
```
Comment (harsimranmaan) [inline lib/widgets/shipanther_scaffold.dart:190]: This is just noise. It should be removed from PRs as it leads to unnecessary cycle.s

File: android/settings_aar.gradle [hunk 5]
```diff
@@ -0,0 +1 @@
+include ':app'
```
Comment (harsimranmaan) [inline android/settings_aar.gradle:1]: Not sure where this is coming from but this should likely be ignored

File: lib/l10n/shipanther_localization.dart [hunk 6]
```diff
@@ -174,6 +174,36 @@ class ShipantherLocalizations {
...
+        locale: locale.toString(),
+      );
+  String get passwordUpdateError => Intl.message(
+        'Password does not match',
+        name: 'passwordUpdateError',
+        desc: 'Password does not match',
+        locale: locale.toString(),
+      );
+  String get username => Intl.message(
+        'Username: ',
```
Comment (harsimranmaan) [inline lib/l10n/shipanther_localization.dart:196]: Why use the `:`. We don't use it anywhere. Text should be consistent across the board

**Comment by harsimranmaan** [comment]
I think that the password change flow should ask and confirm old password only then allow the user to change the password

File: lib/screens/profile.dart [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+                                  labelText:
+                                      ShipantherLocalizations.of(context).name),
+                              autocorrect: false,
+                              enableSuggestions: false,
+                              keyboardType: TextInputType.text,
+                              validator: (String value) {
+                                if (value.isEmpty) {
+                                  return ShipantherLocalizations.of(context)
+                                      .paramRequired(
+                                          ShipantherLocalizations.of(context)
```
Comment (harsimranmaan) [inline lib/screens/profile.dart:93]: Shouldn't this say that username is required?

File: lib/screens/profile.dart [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+                              onSaved: (val) =>
+                                  setState(() => _updatedName = val),
+                            ),
+                            Container(
+                              padding:
+                                  const EdgeInsets.symmetric(vertical: 16.0),
+                              alignment: Alignment.center,
+                              child: SignInButtonBuilder(
+                                icon: Icons.save,
+                                backgroundColor: Colors.blue,
```
Comment (harsimranmaan) [inline lib/screens/profile.dart:108]: Buttons should follow a consistent theme throughout the app. We don't want to put colors all over the place

**Comment by harsimranmaan** [comment]
Re-login is a requirement for password setup.
```
flutter: Unhandled error [firebase_auth/requires-recent-login] This operation is sensitive and requires recent authentication. Log in again before retrying this request. occurred in Instance of 'AuthBloc'.
```

File: lib/bloc/auth/auth_bloc.dart [hunk 9]
```diff
@@ -79,5 +79,17 @@ class AuthBloc extends Bloc<AuthEvent, AuthState> {
...
+    if (event is UpdatePassword) {
+      var user = _authRepository.loggedInUser();
+      // TODO: Validate old password here
+      await user.updatePassword(event.newPassword);
+
+      yield AuthInitial();
+    }
+    if (event is UpdateName) {
+      var user = _authRepository.loggedInUser();
+      await user.updateProfile(displayName: event.name);
```
Comment (harsimranmaan) [inline lib/bloc/auth/auth_bloc.dart:91]: This does not update the username for the backend. You'd need to update the user model from the backend.

**Comment by harsimranmaan** [comment]
I am merging this now but this is still not finished. The old password should be validated before change and the user name change should be on the api user model. Once the other important ui changes are done, this should be revisited next week @anshul2401

**Review by harsimranmaan — APPROVED** [review]

**Review by harsimranmaan — APPROVED** [review]

**Pull request merged**